File: leetcode/arrays/121-best-time-to-buy-and-sell-stock.js
// 121. Best Time to Buy and Sell Stock
// You are given an array prices where prices[i] is the price of a given stock on the ith day.
// You want to maximize your profit by choosing a single day to buy one stock and choosing a different day in the future to sell that stock.
// Return the maximum profit you can achieve from this transaction.If you cannot achieve any profit, return 0.

// Example 1:
// Input: prices = [7, 1, 5, 3, 6, 4]
// Output: 5
// Explanation: Buy on day 2(price = 1) and sell on day 5(price = 6), profit = 6 - 1 = 5.
// Note that buying on day 2 and selling on day 1 is not allowed because you must buy before you sell.

//   Example 2:
// Input: prices = [7, 6, 4, 3, 1]
// Output: 0
// Explanation: In this case, no transactions are done and the max profit = 0.

// Constraints:
// 1 <= prices.length <= 105
// 0 <= prices[i] <= 104

//  @param {number[]} prices
//  @return {number}

// solution (not my solution) 5/24/22

const maxProfit = (prices) => {
  let profit = 0;
  let min = prices[0];

  for (let i = 0; i < prices.length; i++) {
    if (min > prices[i]) {
      min = prices[i];
    } else if (prices[i] - min > profit) {
      profit = prices[i] - min;
    }
  }
  return profit;
};

/* 7/6/22
sliding window
initialize profit at 0 because return 0 if no profit achieveable
intiialize min at prices[0]

for loop, iterate through prices and check min and profit need to be reassigned
  check if min is less than prices[i], if true reassign min
  ELSE: check if current profit (prices[i] - min) is greater than existing profit, if true reassign profit

  exit loop
  return profit
 */

const maxProfit = (prices) => {
  let profit = 0;
  let min = prices[0];

  for (let i = 0; i < prices.length; i++) {
    if (min > prices[i]) {
      min = prices[i];
    } else if (profit < prices[i] - min) {
      profit = prices[i] - min;
    }
  }

  return profit;
};

/* 7/6/22 part 2
return max profit or 0
initialize profit at 0
initialize min at prices[0]

loop through prices
  compare against min, if el < min, reassign min
  ELSE: compare profit against el - min, if profit is less, reassign profit

exit loop
return profit
 */
const maxProfit = (prices) => {
  let profit = 0;
  let min = prices[0];

  for (let i = 0; i < prices.length; i++) {
    if (min > prices[i]) {
      min = prices[i];
    } else if (profit < prices[i] - min) {
      profit = prices[i] - min;
    }
  }
  return profit;
};

File: leetcode/arrays/121-best-time-to-buy-and-sell-stock.py
# track both min and profit
# return max profit

# initialize left and right pointers at index 0 and 1 (left is min, right is max)
# initialize maxP at 0

# interate while r is less than len(prices)
#   check if profitable - price at l is less than price at r
#     true - calc profit and compare against maxP
#     else - assign l to r bc r is the minimum

#   move r 1 index to the right

# return maxP

class Solution:
  def maxProfit(self, prices: [int]) -> int:
    l, r = 0, 1 # l is buy, r is sell
    maxP = 0

    while r < len(prices):
      #profitable
      if prices[l] < prices[r]:
        profit = prices[r] - prices[l]
        maxP = max(maxP, profit)
      #not profitable - move l pointer to new min, which is currently r
      else:
        l = r
      r += 1
    return maxP


class Solution:
  def maxProfit(self, prices: [int]) -> int:
    l, r = 0, 1 # l = min, r = max
    maxP = 0

    while r < len(prices):
      # profitable?
      if prices[l] < prices[r]:
        profit = prices[r] - prices[l]
        maxP = max(maxP, profit)
      #not profitable
      else:
        l = r #the right is lower than left, so move left to lowest
      r += 1

    return maxP
